src/Microsoft.ML.Data/Transforms/HashTransform.cs

```diff
@@ -188,109 +199,65 @@ private static VersionInfo GetVersionInfo()
+                throw Host.ExceptParam(nameof(inputSchema), reason);
         }
-
-        public override void Save(ModelSaveContext ctx)
+        private static (string input, string output)[] GetColumnPairs(ColumnInfo[] columns)
         {
-            Host.CheckValue(ctx, nameof(ctx));
-            ctx.CheckAtModel();
-            ctx.SetVersionInfo(GetVersionInfo());
-
-            // <prefix handled in static Create method>
-            // <base>
-            // Exes
-
-            SaveBase(ctx);
-            Host.Assert(_exes.Length == Infos.Length);
-            for (int iinfo = 0; iinfo < Infos.Length; iinfo++)
-                _exes[iinfo].Save(ctx);
-
-            TextModelHelper.SaveAll(Host, ctx, Infos.Length, _keyValues);
+            Contracts.CheckValue(columns, nameof(columns));
+            return columns.Select(x => (x.Input, x.Output)).ToArray();
         }
 
-        /// <summary>
-        /// Convenience constructor for public facing API.
-        /// </summary>
-        /// <param name="env">Host Environment.</param>
-        /// <param name="input">Input <see cref="IDataView"/>. This is the output from previous transform or loader.</param>
-        /// <param name="name">Name of the output column.</param>
-        /// <param name="source">Name of the column to be transformed. If this is null '<paramref name="name"/>' will be used.</param>
-        /// <param name="hashBits">Number of bits to hash into. Must be between 1 and 31, inclusive.</param>
-        /// <param name="invertHash">Limit the number of keys used to generate the slot name to this many. 0 means no invert hashing, -1 means no limit.</param>
-        public HashTransform(IHostEnvironment env,
-            IDataView input,
-            string name,
-            string source = null,
-            int hashBits = Defaults.HashBits,
-            int invertHash = Defaults.InvertHash)
-            : this(env, new Arguments() {
-                Column = new[] { new Column() { Source = source ?? name, Name = name } },
-                HashBits = hashBits, InvertHash = invertHash }, input)
+        internal ColumnType GetOutputType(ISchema inputSchema, ColumnInfo column)
         {
+            var keyCount = column.HashBits < 31 ? 1 << column.HashBits : 0;
+            inputSchema.TryGetColumnIndex(column.Input, out int srcCol);
+            var itemType = new KeyType(DataKind.U4, 0, keyCount, keyCount > 0);
+            var srcType = inputSchema.GetColumnType(srcCol);
+            if (!srcType.IsVector)
+                return itemType;
+            else
+                return new VectorType(itemType, srcType.VectorSize);
         }
 
-        public HashTransform(IHostEnvironment env, Arguments args, IDataView input)
-            : base(Contracts.CheckRef(env, nameof(env)), RegistrationName, env.CheckRef(args, nameof(args)).Column,
-                input, TestType)
+        public HashTransform(IHostEnvironment env, IDataView input, ColumnInfo[] columns) :
```

Comment:
>HashTransform [](start = 15, length = 13)

can I have a convenience ctor with only one column pair?

Reply:
You have it in estimator, why you want it in transformer?

---
In reply to: [218907955](https://github.com/dotnet/machinelearning/pull/944#discussion_r218907955) [](ancestors = 218907955)

Reply:
Actually, I want users to be able to create transformer directly with no `input`.
Can we have two ctors? `internal` one that you have now, and also `public HashTransform(env, columns)` that checks that `invertHash` is not used in any of the columns?

---
In reply to: [218909081](https://github.com/dotnet/machinelearning/pull/944#discussion_r218909081) [](ancestors = 218909081,218907955)